File: regions/system/library/python/featureflag/src/k1s0_featureflag/memory.py
"""InMemoryFeatureFlagClient 実装"""

from __future__ import annotations

from .exceptions import FeatureFlagError, FeatureFlagErrorCodes
from .models import EvaluationContext, EvaluationResult, FeatureFlag


class InMemoryFeatureFlagClient:
    """テスト用インメモリフィーチャーフラグクライアント。"""

    def __init__(self) -> None:
        self._flags: dict[str, FeatureFlag] = {}

    def set_flag(self, flag: FeatureFlag) -> None:
        """フラグを設定する。"""
        self._flags[flag.flag_key] = flag

    async def evaluate(
        self, flag_key: str, context: EvaluationContext
    ) -> EvaluationResult:
        flag = self._flags.get(flag_key)
        if flag is None:
            raise FeatureFlagError(
                FeatureFlagErrorCodes.FLAG_NOT_FOUND,
                f"フラグが見つかりません: {flag_key}",
            )
        return EvaluationResult(
            flag_key=flag_key,
            enabled=flag.enabled,
            variant=flag.variants[0].name if flag.variants else None,
            reason="FLAG_ENABLED" if flag.enabled else "FLAG_DISABLED",
        )

    async def get_flag(self, flag_key: str) -> FeatureFlag:
        flag = self._flags.get(flag_key)
        if flag is None:
            raise FeatureFlagError(
                FeatureFlagErrorCodes.FLAG_NOT_FOUND,
                f"フラグが見つかりません: {flag_key}",
            )
        return flag

    async def is_enabled(self, flag_key: str, context: EvaluationContext) -> bool:
        result = await self.evaluate(flag_key, context)
        return result.enabled

File: regions/system/library/go/featureflag/memory.go
package featureflag

import (
	"context"
	"sync"
)

// InMemoryFeatureFlagClient はメモリ内フィーチャーフラグクライアントの実装。
type InMemoryFeatureFlagClient struct {
	mu    sync.RWMutex
	flags map[string]*FeatureFlag
}

// NewInMemoryFeatureFlagClient は新しい InMemoryFeatureFlagClient を生成する。
func NewInMemoryFeatureFlagClient() *InMemoryFeatureFlagClient {
	return &InMemoryFeatureFlagClient{
		flags: make(map[string]*FeatureFlag),
	}
}

// SetFlag はフラグを設定する。
func (c *InMemoryFeatureFlagClient) SetFlag(flag *FeatureFlag) {
	c.mu.Lock()
	defer c.mu.Unlock()
	copy := *flag
	c.flags[flag.FlagKey] = &copy
}

func (c *InMemoryFeatureFlagClient) Evaluate(_ context.Context, flagKey string, _ *EvaluationContext) (*EvaluationResult, error) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	flag, ok := c.flags[flagKey]
	if !ok {
		return nil, NewFlagNotFoundError(flagKey)
	}

	result := &EvaluationResult{
		FlagKey: flagKey,
		Enabled: flag.Enabled,
	}
	if flag.Enabled {
		result.Reason = "FLAG_ENABLED"
	} else {
		result.Reason = "FLAG_DISABLED"
	}
	if len(flag.Variants) > 0 {
		v := flag.Variants[0].Name
		result.Variant = &v
	}
	return result, nil
}

func (c *InMemoryFeatureFlagClient) GetFlag(_ context.Context, flagKey string) (*FeatureFlag, error) {
	c.mu.RLock()
	defer c.mu.RUnlock()

	flag, ok := c.flags[flagKey]
	if !ok {
		return nil, NewFlagNotFoundError(flagKey)
	}
	copy := *flag
	return &copy, nil
}

func (c *InMemoryFeatureFlagClient) IsEnabled(ctx context.Context, flagKey string, evalCtx *EvaluationContext) (bool, error) {
	result, err := c.Evaluate(ctx, flagKey, evalCtx)
	if err != nil {
		return false, err
	}
	return result.Enabled, nil
}
